### PR TITLE
Allow build scripts to report error messages through `cargo::error`

### PR DESCRIPTION
**Pull request by torhovland**
Related to #10159.

Adds the `cargo::error` build script directive. It is similar to `cargo::warning`, but it emits an error message and it also fails the build.

### CODE REVIEW

File: tests/testsuite/build_script.rs [hunk 1]
```diff
@@ -3925,14 +3929,16 @@ fn warnings_emitted_when_build_script_panics() {
 }
 
 #[cargo_test]
-fn warnings_hidden_for_upstream() {
+fn errors_and_warnings_hidden_for_upstream() {
```
Comment (epage) [inline tests/testsuite/build_script.rs:3936]: The name of this test implies errors are hidden but they aren't.

In general, I'm feeling we should probably test errors and warnings separately.

File: tests/testsuite/warn_on_failure.rs [hunk 2]
```diff
@@ -90,17 +88,14 @@ fn no_error_or_warning_on_bin_failure() {
...
-
-"#]])
+        .with_stderr_does_not_contain("hidden stderr")
+        .with_stderr_does_not_contain(&format!("[ERROR] {}", ERROR))
+        .with_stderr_does_not_contain(&format!("[WARNING] {}", WARNING1))
+        .with_stderr_does_not_contain(&format!("[WARNING] {}", WARNING2))
+        .with_stderr_contains("[UPDATING] `[..]` index")
+        .with_stderr_contains("[DOWNLOADED] bar v0.0.1 ([..])")
+        .with_stderr_contains("[COMPILING] bar v0.0.1")
+        .with_stderr_contains("[COMPILING] foo v0.0.1 ([..])")
```
Comment (epage) [inline tests/testsuite/warn_on_failure.rs:98]: Why was the test style changed?  We are trying to move away from `does_not_contain` and `contains`

File: tests/testsuite/warn_on_failure.rs [hunk 3]
```diff
@@ -113,16 +108,13 @@ fn errors_and_warnings_on_lib_failure() {
...
-[LOCKING] 2 packages to latest compatible versions
-[DOWNLOADING] crates ...
-[DOWNLOADED] bar v0.0.1 (registry `dummy-registry`)
-[COMPILING] bar v0.0.1
-[ERROR] invalid output in build script of `bar v0.0.1`: `cargo::error=Hello! I'm an error. :)`
-Unknown key: `error`.
-See https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script for more information about build script outputs.
-
-"#]])
+        .with_stderr_does_not_contain("hidden stderr")
```
Comment (epage) [inline tests/testsuite/warn_on_failure.rs:111]: ditto

File: src/doc/src/reference/build-scripts.md [hunk 4]
```diff
@@ -313,13 +314,18 @@ link-arg=FLAG` option][link-arg] to the compiler, but only when building a
...
+script has finished running. Errors are only shown for `path` dependencies
+(that is, those you're working on locally), so for example errors printed
 out in [crates.io] crates are not emitted by default. The `-vv` "very verbose"
-flag may be used to have Cargo display warnings for all crates.
+flag may be used to have Cargo display errors for all crates.
+
+### `cargo::warning=MESSAGE` {#cargo-warning}
+
+The `warning` instruction is just like [`error`](#cargo-error), except that it
+will output warnings.
```
Comment (epage) [inline src/doc/src/reference/build-scripts.md:328]: This does not match what the cargo team accepted in https://github.com/rust-lang/cargo/issues/10159#issuecomment-988104375

`cargo::error` should force the build to fail just like a bad exit code, independent of where it is in the dependency graph.  This means we should always print a `cargo::error`, even for transitive dependencies.

File: tests/testsuite/build_script.rs (hunk 1 above)
Comment (epage) [reply tests/testsuite/build_script.rs:3936]: Ah, this looks to be part of a larger misunderstanding, see https://github.com/rust-lang/cargo/pull/14435#discussion_r1725100785

File: tests/testsuite/warn_on_failure.rs (hunk 2 above)
Comment (torhovland) [reply tests/testsuite/warn_on_failure.rs:98]: It's actually not changed overall. I upgraded the style in the first commit, but left it as it was in this one, so that the PR as a whole only changes one line. But I can update the styles overall.

File: src/doc/src/reference/build-scripts.md (hunk 4 above)
Comment (torhovland) [reply src/doc/src/reference/build-scripts.md:328]: My thinking was to leave this for now, as there was a discussion about this later in the issue. But I can change it.

File: tests/testsuite/warn_on_failure.rs (hunk 2 above)
Comment (epage) [reply tests/testsuite/warn_on_failure.rs:98]: Ah, it was switched back and forth

File: src/doc/src/reference/build-scripts.md (hunk 4 above)
Comment (torhovland) [reply src/doc/src/reference/build-scripts.md:328]: I don't think a feature flag makes a lot of sense here, as a library that emits `cargo::error` would have to require that any client app runs with the feature flag enabled. I'll go with alternative 1.

**Comment by torhovland** [comment]
@epage This is ready for another look.

File: tests/testsuite/build_script.rs [hunk 5]
```diff
@@ -3871,14 +3871,104 @@ fn warnings_emitted() {
...
+                }
+            "#,
+        )
+        .build();
+
+    p.cargo("build")
+        .with_status(101)
+        .with_stderr_data(str![[r#"
+[COMPILING] foo v0.5.0 ([ROOT]/foo)
+[ERROR] invalid output in build script of `foo v0.5.0 ([ROOT]/foo)`: `cargo::error=foo err`
```
Comment (epage) [inline tests/testsuite/build_script.rs]: nit: to really make the behavior obvious in the test change, this commit could use the `cargo::warning`s instead of `cargo::error`s and then the following commit switches it with the introduction of `cargo::error`

Went ahead and tried this out and it works quite well.  Looks like I don't have permission to edit your branch.  Feel free to pull in https://github.com/epage/cargo/tree/build-script-errors

File: tests/testsuite/build_script.rs [hunk 6]
```diff
@@ -4030,20 +4270,10 @@ fn warnings_printed_on_vv() {
...
+...
 [WARNING] bar@0.1.0: foo
 [WARNING] bar@0.1.0: bar
-[RUNNING] `[..] rustc --crate-name bar [..]`
-[COMPILING] foo v0.5.0 ([ROOT]/foo)
-[RUNNING] `[..] rustc --crate-name foo [..]`
-[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
-
+...
 "#]])
```
Comment (epage) [inline tests/testsuite/build_script.rs:4275]: There a reason this test was changed?  Using `...` makes things not update as cleanly

File: tests/testsuite/warn_on_failure.rs [hunk 7]
```diff
@@ -86,13 +86,23 @@ fn no_warning_on_bin_failure() {
...
+[DOWNLOADED] bar v0.0.1 (registry `dummy-registry`)
+[COMPILING] bar v0.0.1
+[COMPILING] foo v0.0.1 ([ROOT]/foo)
+error[E0425]: cannot find function `hi` in this scope
+ --> src/main.rs:1:13
+  |
+1 | fn main() { hi() }
+  |             ^^ not found in this scope
+
+For more information about this error, try `rustc --explain E0425`.
```
Comment (epage) [inline tests/testsuite/warn_on_failure.rs:102]: We try to use wildcards in place of rustc error messages to make it easier for them to change them

File: tests/testsuite/warn_on_failure.rs [hunk 8]
```diff
@@ -86,13 +86,23 @@ fn no_warning_on_bin_failure() {
...
         .with_status(101)
         .with_stdout_does_not_contain("hidden stdout")
-        .with_stderr_does_not_contain("hidden stderr")
-        .with_stderr_does_not_contain(&format!("[WARNING] {}", WARNING1))
-        .with_stderr_does_not_contain(&format!("[WARNING] {}", WARNING2))
-        .with_stderr_contains("[UPDATING] `[..]` index")
-        .with_stderr_contains("[DOWNLOADED] bar v0.0.1 ([..])")
-        .with_stderr_contains("[COMPILING] bar v0.0.1")
-        .with_stderr_contains("[COMPILING] foo v0.0.1 ([..])")
+        .with_stderr_data(str![[r#"
```
Comment (epage) [inline tests/testsuite/warn_on_failure.rs:89]: Ideally this cleanup would be in its own commit

File: src/doc/src/reference/build-scripts.md [hunk 9]
```diff
@@ -313,13 +314,19 @@ link-arg=FLAG` option][link-arg] to the compiler, but only when building a
...
 ### `cargo::warning=MESSAGE` {#cargo-warning}
 
 The `warning` instruction tells Cargo to display a warning after the build
 script has finished running. Warnings are only shown for `path` dependencies
 (that is, those you're working on locally), so for example warnings printed
-out in [crates.io] crates are not emitted by default. The `-vv` "very verbose"
-flag may be used to have Cargo display warnings for all crates.
+out in [crates.io] crates are not emitted by default, unless the build fails.
+The `-vv` "very verbose" flag may be used to have Cargo display warnings for
+all crates.
```
Comment (epage) [inline src/doc/src/reference/build-scripts.md:329]: This should be its own commit

File: tests/testsuite/build_script.rs (hunk 5 above)
Comment (torhovland) [reply tests/testsuite/build_script.rs]: Done.

File: tests/testsuite/build_script.rs (hunk 6 above)
Comment (torhovland) [reply tests/testsuite/build_script.rs]: Reverted to the old snapshot.

File: tests/testsuite/warn_on_failure.rs (hunk 7 above)
Comment (torhovland) [reply tests/testsuite/warn_on_failure.rs]: Fixed.

File: tests/testsuite/warn_on_failure.rs (hunk 8 above)
Comment (torhovland) [reply tests/testsuite/warn_on_failure.rs:89]: Done.

File: src/doc/src/reference/build-scripts.md (hunk 9 above)
Comment (torhovland) [reply src/doc/src/reference/build-scripts.md:329]: Fixed.

**Comment by torhovland** [comment]
@epage This is ready for another look.

File: src/doc/src/reference/build-scripts.md [hunk 10]
```diff
@@ -313,13 +314,19 @@ link-arg=FLAG` option][link-arg] to the compiler, but only when building a
 `cdylib` library target. Its usage is highly platform specific. It is useful
 to set the shared library version or the runtime-path.
 
+### `cargo::error=MESSAGE` {#cargo-error}
+
+The `error` instruction tells Cargo to display an error after the build script
+has finished running, and then fail the build.
```
[the remark this answers is not in the document]
Comment (torhovland) [reply src/doc/src/reference/build-scripts.md:320]: So... should I commit this?

[the remark this answers is not in the document]
Comment (weihanglo) [reply src/doc/src/reference/build-scripts.md:320]: I guess sk and we can always update it later

File: tests/testsuite/warn_on_failure.rs [hunk 11]
```diff
@@ -95,7 +95,6 @@ fn no_warning_on_bin_failure() {
 [COMPILING] foo v0.0.1 ([ROOT]/foo)
 error[E0425]: cannot find function `hi` in this scope
 ...
-For more information about this error, try `rustc --explain E0425`.
```
Comment (epage) [inline tests/testsuite/warn_on_failure.rs]: Could you squash this into the relevant commit?

File: src/doc/src/reference/build-scripts.md [hunk 12]
```diff
@@ -319,6 +319,8 @@ to set the shared library version or the runtime-path.
 The `error` instruction tells Cargo to display an error after the build script
 has finished running, and then fail the build.
 
+> **MSRV:** Respected as of 1.82
```
Comment (epage) [inline src/doc/src/reference/build-scripts.md:322]: Keeping a comment open on this so we remember to update it if this slips

**Comment by epage** [comment]
@rfcbot fcp merge

@torhovland is proposing we insta-stabilize `cargo::error=` build directive which was requested in #10159.

This is following the semantics proposed in a cargo team meeting (https://github.com/rust-lang/cargo/issues/10159#issuecomment-988104375):  `cargo::error=` forces the build script to fail, independent of whether its local or not

@kornelski counters that this means we have two ways to cause a build script to error when we should only have one.  They suggest warning if an error appears without a failure (https://github.com/rust-lang/cargo/issues/10159#issuecomment-2305139091).  However, that does mean that we have error messages that might not be tied to a failure which can cause user confusion and I think we should keep the Cargo team's previous proposal on this.

This is only supported with `cargo::` syntax and not `cargo:` syntax.  Use of `cargo:error` would create an `error` metadata entry.  As `cargo::` is strictly checked, this will require an MSRV bump to use which is good in this case if people are relying on `cargo::error` causing a bad  exit code.  Because `cargo::error` will fail on previous versions of Cargo, no transition period should be needed for this.

As for insta-stabiliaze, I can understand that it seems like a lot to create an unstable feature just for this build directive.  We don't have anything like `-Zunstable-options` for build directives.  This is relatively small in scope and low level / mechanism focused, that besides @kornelski's counter proposal, I'm not seeing much we can learn from an extended testing period.  I also don't expect much nightly testing on this.

File: tests/testsuite/warn_on_failure.rs (hunk 11 above)
Comment (torhovland) [reply tests/testsuite/warn_on_failure.rs]: Done.

**Comment by kornelski** [comment]
The primary motivation for this feature was #10159, where the main problem was noisy output from `pkg_config`. However, I don't see how to integrate `cargo::error` with the `pkg_config` crate ;(

[By default](https://docs.rs/pkg-config/latest/pkg_config/struct.Config.html#method.cargo_metadata) `pkg_config` prints cargo metadata as a side effect. But this metadata must not contain `cargo::error`, because the API returns a `Result` and many crates try a fallback after that. `pkg_config` can't stuff `cargo::error` prefix in `Display` of `pkg_config::Error` either, because some crates have `printf("cargo:warning={pkg_config_error}");` and that would nest syntax, or `cargo::error` after a newline would fail the build instead of warning.

**Comment by ehuss** [comment]
I would assume that the caller of pkg-config would be responsible for printing `cargo::error` with the appropriate error if they want to. Can you explain why that might not work?

**Comment by kornelski** [comment]
@ehuss The downside is that instead of improving one crate to benefit the whole ecosystem, it'll be necessary to change 750 downstream crates.

 `pkg_config` instructs users to `.unwrap()` it: https://docs.rs/pkg-config/latest/pkg_config/#example

system-deps [too](https://docs.rs/system-deps/latest/system_deps/#usage).

which means these crates can't adopt `cargo:error` on their own, in their current API, to automatically provide a good baseline to all current users who `unwrap()` their errors.

**Comment by epage** [comment]
@torhovland looks like some other changes cause the output in the new tests to change and the commits need to be updated.

**Comment by ehuss** [comment]
Closing in favor of #14743.
